### ISSUE TITLE
[RFC] [Console] Virtual Terminal

### ISSUE THREAD

**Issue by joelwurtz** [symfony/symfony#51407]
### Description

After working some times with the `ConsoleSectionOutput` and `SymfonyStyle` class inside the Console Component i found out that the two of them does not work well together and that even the `ConsoleSectionOutput` does not handle a lot of use case.

As an example asking a question in a section is very random in term of output given the order of code declaration. The same goes for using a title inside section (extra new lines at some points, max height clearing too many lines, etc ...)

I clearly saw that writing something like this is not trivial so big up to the existing persons that have worked on this, it work well for a lot of use cases.

However i think that to fix some of the remaining use case, and add a lot of nice features it need to have a different logic than the current one.

This RFC is mainly to propose a new solution and talk about it before writing too many codes that may be not wanted into the console component.

To better handle section i would like to add a Terminal Emulator system. 

It would act as an OutputInterface and parse the bytes writed to it to decode and handle escaped terminal sequence (like moving up/down/left/right the cursor, writing colored text, etc ....). 
Will doing so it would keep the current state of this terminal (where is the cursor, what is the current text format, what is the visible content with the correct size, the non visible content : so we can handle moving the cursor into a non visible part, etc ....)

There is some sources of inspiration that i have see from other languages : 

 * https://github.com/hinshun/vt10x
 * https://github.com/mlochbaum/st
 * https://github.com/wez/wezterm/tree/main/term

The goal is to provide a base for terminal emulation for the 80% use case at first (cursor and formatting as a base would be nice) other things like mouse support or edge case espace terminal sequence can be added in the future.

There would also be a way to render a terminal into another terminal at a specific position, which would allow to have the same behavior as the `ConsoleSectionOutput` class. 

There would also be helper to automatic size those terminal into the main given there current height (so we can mimic the current `maxHeight` property of section)

I think having this would definitely allow the `ConsoleSectionOutput` to have a better feeling for end user and would fix a lot of edge case. 

### Example

In my mind user would have to write something like this : 

```php
$mainTerminal = Terminal::fromEnv(); // Create a terminal from current environnement
$virtualTerminal1 = new VirtualTerminal(height: 20, cols: 50);
$virtualTerminal2 = new VirtualTerminal(height: 20, cols: 50);
$virtualTerminal3 = new VirtualTerminal(height: 20, cols: 50);

$mainTerminal->attach($virtualTerminal1, x: 0, y: 0);
$mainTerminal->attach($virtualTerminal2, x: 20, y: 0);
$virtualTerminal2->attach($virtualTerminal3, x: 20, y: 0); // Would render starting x at 40 in the main terminal

$virtualTerminal1->write('Foo');
$virtualTerminal2->write('Bar');
$virtualTerminal3->write('Baz');

// This would output
// Foo -> 19 new lines -> Bar -> 19 new lines -> Baz
```

**Comment by lyrixx** [comment]
I like it 👍🏼

**Comment by stof** [comment]
Given that terminals don't allow you to move the cursor higher than the top of the viewport, how you would implement a virtual terminal not having this limitation ? In the end, we still need to render to an actual terminal (otherwise this API is useless)

**Comment by joelwurtz** [comment]
> Given that terminals don't allow you to move the cursor higher than the top of the viewport, how you would implement a virtual terminal not having this limitation

Mainly by buffering the content, but maybe at some point this will give too much memory pressure, so we could add a maximum size for this buffer and throw error / sanitize cursor movement if out of scope

Cursor would be virtual also (which means that each terminal would have it's own cursor state) it does not impact the cursor of the "main" terminal

EDIT: As an example it would give something like

 * virtual terminal as an height of 50 and is attached to the "main" terminal
 * user write 100 lines of content 
 * it display only 50 last lines in the "main" terminal (but the 100 lines are actually buffered)
 * user move cursor up to 60 lines and clean content (not sure in the implementation if it would clear the content in the buffer also or only have a specific state to make them "unrenderable")
 * it display only 40 lines of the content in the "main" terminal

**Comment by stof** [comment]
@joelwurtz writing only the last 50 lines is still bad as it means that the user has no way to see the previous 50 lines when scrolling up their terminal window, breaking the usage of the console (especially for cases where you don't try to overwrite the first 50 lines)

**Comment by joelwurtz** [comment]
This is intended, scrolling may be achieve in the future by listening to mouse event, but the goal is to have something that can handle well rendering separated output from multiple sources (like 2 process in parallel)

In my mind it would be done like this

 * process A -> go to the first virtual terminal
 * process B -> go to the second virtual terminal
 * once a process is terminated, it output everything is the main terminal (where there is history and scroll control) and remove the virtual terminal associated
 
This way we can separate those output and have a clear history in the end

But yes this virtual terminal is not meant to be used everywhere, it's mainly used when you need multiplexing or could also be useful for test purposes

**Comment by stof** [comment]
@joelwurtz I'm talking about the fact that most terminal apps have a scrolling behavior allowing to see what was output before (even if control sequences are not allowed to go outside of the viewport to override things). If you don't output things to the native terminal, this previous output will not exist.

Also, if the goal is to be able to render those virtual terminals side by side, I fear this will become very complex (as control sequences allowing to override things are working at the line level)

**Comment by stof** [comment]
Thus, I'm not sure this will actually work better in practice than the existing sections with max height.

**Comment by joelwurtz** [comment]
> Also, if the goal is to be able to render those virtual terminals side by side, I fear this will become very complex (as control sequences allowing to override things are working at the line level)

That's perfectly why this class is about: reduce complexity of those cases.

When a control sequence will be written to a virtual terminal it will not be output to the real terminal, it will be parsed and analyzed given the current content of this virtual terminal (if a control sequence erase a line it will do so but only in the buffer of the specific virtual terminal). 

Once this is done, the virtual terminal would render different lines to the main terminal (without the original control sequence obivously, but maybe new one to render / overwrite something side by side). 

In fact it's a terminal, it's only "virtual" because it does not render something directly and need a "real" terminal to make something visible

**Comment by stof** [comment]
@joelwurtz I'm fearing the complexity of the VirtualTerminal itself (even more if you expect it to parse control sequences in its input, which is even more complex than how sections work).
To me, the only "benefit" of this virtual terminal over sections with max height is the ability to render sections side by side. But this looks like lots of complexity for a small feature (and even worse to make it cross-platform).
Having the virtual terminal becoming bigger than the actual terminal is something that would not work well IMO as you would loose the native features of the terminal, making things worse for users.

You say that the existing ConsoleSection is too complex to implement it properly, but you suggest to replace it with something even more complex.

**Comment by joelwurtz** [comment]
It's relevant if we want to fix some issues within the `ConsoleSectionOutput` class, actually if we write an escape sequence inside it (which can be easily done if we pass the output of a process to it, like phpstan / docker / ...) it does not work as the section is not aware of this state. And it will never work if we don't parse them.

The question is then : do you want the console section output to support escape sequence ?

If yes, then this RFC is an attempt to normalize this implementation around a concept that exist in other languages. And yes it will be "complex" but at least it will use an existing concept where we can relate to in other libraries to see how they resolve some issues.

**Comment by maxbeckers** [comment]
👍 for optimizing `ConsoleSectionOutput`

**Comment by carsonbot** [comment]
Thank you for this suggestion.
There has not been a lot of activity here for a while. Would you still like to see this feature?

**Comment by joelwurtz** [comment]
Not stalled, i'm currently waiting for symfony/terminal as it will be certainly the place where this feature should happen

**Comment by carsonbot** [comment]
Thank you for this suggestion.
There has not been a lot of activity here for a while. Would you still like to see this feature?

**Comment by alamirault** [comment]
I'm very interested by this issue, is there any news for `symfony/terminal` ?

**Comment by MatTheCat** [comment]
#63778 has been merged!

**Comment by chalasr** [comment]
Please comment if this needs work so that we reopen
